### ISSUE TITLE
Strange behavior with :terminal resizing

### ISSUE THREAD

**Issue by ColinKennedy** [neovim/neovim#27561]
### Problem

Resizing a terminal buffer alters lines in unexpected ways. This occurs when resizing a terminal via a split or when resizing the entire terminal

See video demo, below:

https://github.com/neovim/neovim/assets/10103049/af03f6ec-ddcd-415d-acce-b47eda25913b

### Steps to reproduce

- Install Neovim nightly (NVIM v0.10.0-dev-2396+gb4c013c27-dirty, in my case)
- Call `nvim -u NONE`
- Run `:terminal`
- Print something with long lines and then resize the terminal

### Expected behavior

Resizing would work without causing blank space which permanently alters the lines

### Neovim version (nvim -v)

v0.10.0-dev-2396+gb4c013c27-dirty

### Vim (not Nvim) behaves the same?

yes, vim 9.1

### Operating system/version

CentOS Linux 7 (Core)

### Terminal name/version

gnome or Alacritty (both have the issue)

### $TERM environment variable

xterm-256color

### Installation

build from repo

**Comment by vurentjie** [comment]
One thing I found that might improves that issue with text truncating is to use a tmux session in the nvim terminal. It's not perfect but it seems to be more forgiving when windows resize. 

Maybe I am imagining things, but I have a demo here you can try: 
https://gist.github.com/vurentjie/a2f878053e977e35afc5c57ff65f4949 
What I do is I create a separate tmux session for each terminal I open in neovim.

**Comment by justinmk** [comment]
Is this different than https://github.com/neovim/neovim/issues/2514 ?

**Comment by ColinKennedy** [comment]
I was under the impression that https://github.com/neovim/neovim/issues/2514 was implemented as Neovim 0.10+'s terminal reflow which would handle this case. In which case this seemed to be unintended behavior of the reflow. It'd be nice to know how to reset the view manually because when it breaks like this, the text stays like that and the only recourse is to run the command again. Terminals are a nebulous subject so I could be wrong on where the bug lies exactly but I can say tmux wasn't involved in this but is reproducible there as well.

**Comment by clason** [comment]
You can try locally reverting the reflow change and see if it makes a difference.

**Comment by ColinKennedy** [comment]
Not at a machine right now to test but I believe it would still be the same because I recall it also happening in Vim 9.1

**Comment by miduddin** [comment]
Not sure if I should open a new issue, but I'm noticing similar problem on my machine:

https://github.com/user-attachments/assets/b354fb63-3022-405c-a598-92c8ef31df32

I'm using WSL + Windows Terminal. In the recording I split the Windows Terminal pane to the right and then close it again to simulate resize, after that as you can see the whole lazygit UI is shifted to the left.

Unlike OP's problem however, Vim works fine for me:

https://github.com/user-attachments/assets/ed721298-6331-4faa-9f9b-49a09af8c487

**Comment by Quaddroo** [comment]
I have an out-of-vim solution for people while a fix brews: run scroll, the 'suckless' utility, in the background for that terminal. This is also (perhaps more-so) applicable to https://github.com/neovim/neovim/issues/2514 (but I can't post it there, maybe someone can link this) . In that issue, it is also mentioned that you may use TMUX in the pane, but that seems very heavy just to fix this bug, and would impede other things, like the scrollback buffer provided by neovim. As far as I can tell, this works very well.

**Comment by isti115** [comment]
I'm experiencing the same offset glitch with `lazygit` as @miduddin and can confirm that the issue is still present in `v0.10.4`. On the other hand, I can't seem to be able to reproduce it using `v0.11.0-dev-1341+g5eda7aafe9`!

**Comment by justinmk** [comment]
> On the other hand, I can't seem to be able to reproduce it using `v0.11.0-dev-1341+g5eda7aafe9`!

Closing then, unless someone confirms it's still present on latest master (mention the exact sha that you tested).

**Comment by ColinKennedy** [comment]
This still happens on master 48e6147e64b819d2a4130697bb348dcd368a4391

(Unlike the original post this is in Wezterm + WSL)

https://github.com/user-attachments/assets/0c4801ef-f0ca-4a1b-9adc-b6a85534024c

That said resizing in general can get funky. It's hard for me to know if Neovim is reflowing properly and the terminal is broken underneath in some way or what.

https://github.com/user-attachments/assets/0a2c4ae8-2446-489b-b0e7-bd48f5812614

**Comment by PeterCardenas** [comment]
@Quaddroo does `scroll lazygit`  work for you on v0.11.2? or are you doing something else

**Comment by Quaddroo** [comment]
I'm not familiar with lazygit. Currently, on neovim v0.11.2, I can still get :terminal oddities after playing around with generating text on the terminal and resizing. Unfortunately, it seems scroll (literally, running "scroll" in the :terminal ) does not help the issue anymore as well. tmux still seems to help immensely, although I did manage to find a few very tiny artifacts.

**Comment by PeterCardenas** [comment]
i have found somewhat of a hack that has worked for me pretty consistently, and maybe might inform the solution:
```lua
vim.api.nvim_create_autocmd('TermOpen', {
  callback = function(args)
    local dirty_buf_enter = false
    local bufnr = args.buf
    -- Triggers the following sequence of events after resizing or opening the terminal buffer:
    -- Resize the terminal to 0 width and 0 height
    -- Wait 50ms
    -- Resize the terminal to 100 width and 100 height
    -- Switch to a temporary buffer and switch back
    -- Resize the terminal to 0 width and 0 height
    -- Wait 50ms
    -- Resize the terminal to 100 width and 100 height
    -- Terminal looks good
    local function correct_size()
      local temp_bufnr = vim.api.nvim_create_buf(false, true)
      vim.cmd('resize 0 0')
      local cur_dirty_buf_enter = dirty_buf_enter
      vim.defer_fn(function()
        vim.cmd('resize 100 100')
        if not cur_dirty_buf_enter then
          dirty_buf_enter = true
          vim.api.nvim_set_current_buf(temp_bufnr)
          vim.api.nvim_set_current_buf(bufnr)
          vim.api.nvim_buf_delete(temp_bufnr, { force = true })
        end
      end, 50)
      if dirty_buf_enter then
        dirty_buf_enter = false
      end
    end
    vim.api.nvim_create_autocmd('BufEnter', {
      buffer = bufnr,
      callback = function()
        correct_size()
      end,
    })
    vim.api.nvim_create_autocmd('VimResized', {
      buffer = bufnr,
      callback = function()
        correct_size()
      end,
    })
  end,
})
```

**Comment by ColinKennedy** [comment]
@PeterCardenas  I'm not sure exactly what this block is meant to do but FWIW I loaded Neovim `NVIM v0.12.0-dev-195+g82f08f33c1` with `nvim -u NONE`, loaded the file with `so resize_fix.lua`, then made a split terminal and it still has the resizing issue, even if I leave the terminal buffer and enter it or resize Vim.

![Image](https://github.com/user-attachments/assets/c34f4bc8-2de6-4a20-9fec-2c9a2546e265)

**Comment by clason** [comment]
Please always test with `nvim --clean` -- that is the (only) intended way of testing Nvim with "factory defaults". (Vim has slightly different behavior with these options; classical instructions for Vim thus don't apply here.)

**Comment by PeterCardenas** [comment]
hmm i guess it's because lazygit handles sigwinch or something (my use case), but the reflow is separate

**Comment by ColinKennedy** [comment]
@clason Thank you for letting me know, I didn't know that. In any case, even with `--clean` it still happens with `nvim --clean`

**Comment by rokonio** [comment]
Can confirm that it is happening even when resizing the terminal from inside of neovim

https://github.com/user-attachments/assets/71015ed3-2836-4c46-a97e-3be95d6b0764

**Comment by cjsftnt** [comment]
Is this because the reflow only works on screen text? when terminal width is reduced, some text is pushed out and when the width is bigger, these text show up but clipped.

**Comment by clason** [comment]
Yes, reflow does _not_ apply to Neovim's internal scrollback buffer, only to what vterm handles itself. This is documented.

**Comment by justinmk** [comment]
Assuming https://github.com/vim/vim/pull/8365 solves terminal reflow, hopefully we can merge that
